### ISSUE TITLE
ValueError: Specs passed to a DirectoryLayout must be concrete!

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#2911]
I've never used modules before, but ANL's LCRC is looking to make the switch to Lmod for our next cluster. I decided to try out our modules tutorial, but ran into a problem. I created a `~/.spack/modules.yaml` containing:
```yaml
modules:
  enable::
    - lmod
```
and ran:
```
$ spack module refresh --module-type lmod
```
but it crashed with:
```
Traceback (most recent call last):
  File "/soft/spack-0.10.0/bin/spack", line 212, in <module>
    main(sys.argv)
  File "/soft/spack-0.10.0/bin/spack", line 208, in main
    _main(args, unknown)
  File "/soft/spack-0.10.0/bin/spack", line 174, in _main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/module.py", line 259, in module
    callbacks[args.subparser_name](module_type, specs, args)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/cmd/module.py", line 243, in refresh
    x.write(overwrite=True)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/modules.py", line 380, in write
    set_module_variables_for_package(package, mod)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/build_environment.py", line 380, in set_module_variables_for_package
    m.std_cmake_args = spack.CMakePackage._std_args(pkg)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/build_systems/cmake.py", line 95, in _std_args
    rpaths = ':'.join(spack.build_environment.get_rpaths(pkg))
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/build_environment.py", line 424, in get_rpaths
    rpaths.extend(d.prefix.lib for d in deps
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/build_environment.py", line 425, in <genexpr>
    if os.path.isdir(d.prefix.lib))
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/spec.py", line 1156, in prefix
    return Prefix(spack.store.layout.path_for_spec(self))
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/directory_layout.py", line 124, in path_for_spec
    _check_concrete(spec)
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/directory_layout.py", line 42, in _check_concrete
    raise ValueError('Specs passed to a DirectoryLayout must be concrete!')
ValueError: Specs passed to a DirectoryLayout must be concrete!
```
This seems to be a common error message. I'm seeing it in #206, #217, #238, #244, and #1960. From what I gather from reading those other issues, it sounds like this is caused by changing hashes and packages/variants being added/removed? Unfortunately, this makes Spack + Modules a pipe dream, as you would have to completely reinstall Spack or uninstall all previous versions the first time any conflict occurs.

**Comment by alalazo** [comment]
> Unfortunately, this makes Spack + Modules a pipe dream, as you would have to completely reinstall Spack or uninstall all previous versions the first time any conflict occurs.

@adamjstewart Does this happen if you use a fresh checkout of `spack`?

**Comment by adamjstewart** [comment]
> Does this happen if you use a fresh checkout of spack?

Nope. I'm assuming that this is because none of the hashes have changed since I checked it out and installed a few packages.

**Comment by alalazo** [comment]
@adamjstewart I think you are right. In my opinion Spack should work consistently *at a given version*. Constantly updating a checkout of `develop` is very likely to be causing these kind of problems and I am not really sure we want to support this use case. I understand this was an hard decision to make before because the releases were not so frequent, but apparently this is going to change. @tgamblin ?

**Comment by alalazo** [comment]
@adamjstewart Anyhow, I took the issue. I'll try to see as soon as I'll have time (sorry, busy week) if I can find a way to reproduce it.

**Comment by adamjstewart** [comment]
Steps to reproduce this error from a fresh clone of Spack:

1. `spack install m4`
2. `spack edit libsigsegv` and add `variant('fake')`
3. `spack install m4`
4. `spack module refresh`

**Comment by adamjstewart** [comment]
> In my opinion Spack should work consistently at a given version. Constantly updating a checkout of develop is very likely to be causing these kind of problems and I am not really sure we want to support this use case. I understand this was an hard decision to make before because the releases were not so frequent, but apparently this is going to change.

I'm not sure that this is realistic. What you are telling me is that if a user asks me for a new package, or asks me to add some additional support to a package, I tell them "sure, but you'll have to wait until the next Spack release before I can install it for you".

**Comment by alalazo** [comment]
The problem here seems to be that with the new variant added for `libsigsegv` the previously installed `m4` is not seen as concrete (very likely because there's no value for `fake` in the previous `libsigsev`).

**Comment by alalazo** [comment]
@adamjstewart I'll try my best to have a look at this tomorrow :smile:

**Comment by tgamblin** [comment]
@adamjstewart: I think we should do our best to work across versions.

**Comment by adamjstewart** [comment]
Thanks @alalazo! I've talked to @tgamblin before about the changing hashes problem and he assured me that I shouldn't have to reinstall Spack every time a variant is added or the hashing algorithm changes. I might end up with hundreds of duplicate installations (currently pushing 1,500 across 3 Spack installations), but that they shouldn't cause any problems. So I'm hoping this is just a "bug" and not a "feature".

**Comment by tgamblin** [comment]
@alalazo:
> The problem here seems to be that with the new variant added for libsigsegv the previously installed m4 is not seen as concrete (very likely because there's no value for fake in the previous libsigsev).

Specs written to the DB are explicitly marked concrete, and they're explicitly read in as concrete as well. Concrete specs cache both the fact that they are concrete (in `self._concrete`) and the hash they were created with (in `self._hash`) specifically so that we respect the hash and package configuration *from the time the package was built*.  This is done specifically to address the issue of changing packages: if a spec is marked concrete, you know it has values for *all* variants *that existed when it was built*.  A missing variant on a concrete spec means the package didn't know about that variant, and a spec without variant `foo` won't satisfy *either* `+foo` or `~foo`. 

So basically I don't think it's because Spack is failing to recognize that a previously installed spec is concrete.  If it is, it's a bug.

**Comment by tgamblin** [comment]
@adamjstewart: 
> From what I gather from reading those other issues, it sounds like this is caused by changing hashes and packages/variants being added/removed?

Spack has a `DirectoryLayout` class that controls how packages are laid out under the install root (usually `opt/spack`).  Among other things, the directory layout guarantees you a unique, deterministically knowable directory for every installed package.  This error is raised when you ask it for a directory corresponding to a spec that is not concrete -- i.e. a spec that could represent more than one unique configuration of a package.

In 2015, Spack didn't do a very good job of managing changing package configurations.  If you added a new variant or a new dependency to a package, Spack would not recognize that old versions of a package were concrete, because the new DAG was a superset of the old one.  The PRs you linked to fixed that issue -- now we remember when an old build was concrete and new builds will have different hashes from old ones, but you can still compare them to each other without getting issues.

**Comment by tgamblin** [comment]
@alalazo: does the module infrastructure try to rebuild or reconcretize old specs rather than sticking with the ones read in from the DB?  If it does, then you can run into this problem.

**Comment by alalazo** [comment]
> does the module infrastructure try to rebuild or reconcretize old specs rather than sticking with the ones read in from the DB?

 I am checking that right now.

**Comment by adamjstewart** [comment]
Possibly useful debugging information.

#### Things that do not cause a problem

```
$ spack install libsigsegv
$ spack edit libsigsegv  # add variant
$ spack install libsigsegv
$ spack module refresh
```
```
$ spack install libsigsegv
$ spack install m4
$ spack edit libsigsegv  # add variant
$ spack install libsigsegv
$ spack install m4
$ spack module refresh
```

#### Things that do cause a problem
```
$ spack install m4
$ spack edit libsigsegv  # add variant
$ spack install m4
$ spack module refresh
```
It seems to make a difference whether or not the package was explicitly installed.

**Comment by alalazo** [comment]
@tgamblin I am investigating where this is coming from:
```python
In[13]: print pkg.spec
m4@1.4.18%gcc@4.8+sigsegv arch=linux-ubuntu14-x86_64 ^libsigsegv@2.10%gcc@4.8 arch=linux-ubuntu14-x86_64
In[14]: pkg.spec.concrete
Out[15]: True
In[15]: pkg.spec['libsigsegv'].concrete
Out[16]: False
```
Debugging objects right before the exception is raised. The `m4` package is clearly inconsistent there.

**Comment by alalazo** [comment]
> So basically I don't think it's because Spack is failing to recognize that a previously installed spec is concrete. If it is, it's a bug.

@tgamblin I think it's a bug:
```
a = specs[1]
a.concrete, a['libsigsegv'].concrete
Out[3]: (True, True)
a = specs[1].package.spec
a.concrete, a['libsigsegv'].concrete
Out[5]: (True, False)
```
In the scope of the main `module` command, before even entering the refresh subcommand. It seems that `spec is not spec.package.spec`.

**Comment by alalazo** [comment]
Ok, the problem seems to be inherent in spec construction from records in the DB. If there's no rush I'll try to put together a PR tomorrow.

**Comment by alalazo** [comment]
Going with a debugger through the failing case provided by @adamjstewart I see that `spec.package` is **always** read from the current repository of packages. This seems to be inconsistent to me. As we store the `package.py` files used to build a particular DAG in the `.spack` subfolder of the prefix I think we should read the packages from there if we retrieve something from a DB. Do you agree @tgamblin or am I overlooking something?

**Comment by alalazo** [comment]
@adamjstewart Can you confirm that #2920 fixes the issue also for you?

**Comment by adamjstewart** [comment]
Saw this error message again today in an unrelated context. I tried deactivating and uninstalling some packages on the latest develop branch. I must have installed some of them on a branch with different packages/variants.

**Comment by davydden** [comment]
also came across the issue today. I wiped out all packages but `gcc` and dependencies, install whole lot of things (`dealii` and freinds) and on module refresh got this error. Probably something changed in variants/dependencies of gcc since i installed it (January 16).

**Comment by pramodk** [comment]
same issue here !
Is there any way to find out which package causing the issue? (instead of removing everything )

**Comment by alalazo** [comment]
@davydden @pramodk If you didn't install a package that does not exist in your current repo, #2920 is a workaround. I'll try to work on a permanent fix along the lines discussed in the PR as soon as possible.

**Comment by davydden** [comment]
> #2920 is a workaround.

we need to have it merged :smile:

**Comment by pramodk** [comment]
thanks @alalazo, #2920 fixed my issue!

**Comment by adamjstewart** [comment]
#2920 was reverted in #3078 so this bug is still present. I am unable to run `spack module refresh` on the current develop branch.

**Comment by alalazo** [comment]
For the record, the same behavior is encountered when uninstalling:
```console
... lot of output omitted
==> Successfully uninstalled isl@0.18%gcc@4.8 arch=linux-ubuntu14-x86_64 /afi6taq
==> Successfully uninstalled libelf@0.8.13%gcc@6.3.0 arch=linux-ubuntu14-x86_64 /7gdliwy
==> Successfully uninstalled libsigsegv@2.10%gcc@4.8 arch=linux-ubuntu14-x86_64 /zyc4d7e
==> Successfully uninstalled libsigsegv@2.10%gcc@6.3.0 arch=linux-ubuntu14-x86_64 /5o7otdp
==> Successfully uninstalled libxml2@2.9.4%gcc@4.8~python arch=linux-ubuntu14-x86_64 /nistiwc
Traceback (most recent call last):
  File "/home/mculpo/PycharmProjects/spack/bin/spack", line 212, in <module>
    main(sys.argv)
  File "/home/mculpo/PycharmProjects/spack/bin/spack", line 208, in main
    _main(args, unknown)
  File "/home/mculpo/PycharmProjects/spack/bin/spack", line 174, in _main
    return_val = command(parser, args)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/cmd/uninstall.py", line 218, in uninstall
    do_uninstall(uninstall_list, args.force)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/cmd/uninstall.py", line 162, in do_uninstall
    item.do_uninstall(force=force)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/package.py", line 1520, in do_uninstall
    spack.hooks.pre_uninstall(self)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/hooks/__init__.py", line 72, in __call__
    hook(*args, **kwargs)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/hooks/extensions.py", line 31, in pre_uninstall
    if pkg.activated:
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/package.py", line 846, in activated
    exts = spack.store.layout.extension_map(self.extendee_spec)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/directory_layout.py", line 362, in extension_map
    _check_concrete(spec)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/directory_layout.py", line 42, in _check_concrete
    raise ValueError('Specs passed to a DirectoryLayout must be concrete!')
ValueError: Specs passed to a DirectoryLayout must be concrete!
```
I think I'll try to fix this after I finish with #3183 (if nobody else is already working on this).

**Comment by alalazo** [comment]
Repeating the actions in https://github.com/spack/spack/issues/2911#issuecomment-274620083 I get:
```console
k$ spack -d module refresh
==> READ LOCK: /home/mculpo/PycharmProjects/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> READ LOCK: /home/mculpo/PycharmProjects/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> READ LOCK: /home/mculpo/PycharmProjects/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> You are about to regenerate tcl module files for:

-- linux-ubuntu14.04-x86_64 / gcc@4.8 ---------------------------
fotodb6 libsigsegv@2.11  25bjtdn libsigsegv@2.11  r26rlmg m4@1.4.18  4huasig m4@1.4.18

==> Do you want to proceed? [y/n] y
==> Regenerating tcl module files
==> 	WRITE: libsigsegv@2.11%gcc@4.8 arch=linux-ubuntu14.04-x86_64 /fotodb6 [/home/mculpo/PycharmProjects/spack/share/spack/modules/linux-ubuntu14.04-x86_64/libsigsegv-2.11-gcc-4.8-fotodb6]
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/compilers.yaml
==> Reading config file /home/mculpo/.spack/linux/compilers.yaml
==> 	WRITE: libsigsegv@2.11%gcc@4.8 fake= arch=linux-ubuntu14.04-x86_64 /25bjtdn [/home/mculpo/PycharmProjects/spack/share/spack/modules/linux-ubuntu14.04-x86_64/libsigsegv-2.11-gcc-4.8-25bjtdn]
==> 	WRITE: m4@1.4.18%gcc@4.8 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=linux-ubuntu14.04-x86_64 /r26rlmg [/home/mculpo/PycharmProjects/spack/share/spack/modules/linux-ubuntu14.04-x86_64/m4-1.4.18-gcc-4.8-r26rlmg]
==> 	WRITE: m4@1.4.18%gcc@4.8 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=linux-ubuntu14.04-x86_64 /4huasig [/home/mculpo/PycharmProjects/spack/share/spack/modules/linux-ubuntu14.04-x86_64/m4-1.4.18-gcc-4.8-4huasig]
```
Closing.
